src/frame/mod.rs

```diff
@@ -54,29 +75,53 @@ impl fmt::Display for Error {
 
 impl std::error::Error for Error {}
 
-/// Compress all bytes of `input`.
-pub fn compress(input: &[u8]) -> Vec<u8> {
+/// Compress all bytes of `input` into a `Vec`.
+pub fn compress(input: &[u8]) -> Result<Vec<u8>, Error> {
     compress_with(FrameInfo::default(), input)
 }
 
-/// Compress all bytes of `input`.
-pub fn compress_with(frame_info: FrameInfo, input: &[u8]) -> Vec<u8> {
+/// Compress all bytes of `input` into a `Vec` with the specified Frame configuration.
+pub fn compress_with(frame_info: FrameInfo, input: &[u8]) -> Result<Vec<u8>, Error> {
     let buffer = Vec::with_capacity(
         header::MAX_FRAME_INFO_SIZE
             + header::BLOCK_INFO_SIZE
             + crate::block::compress::get_maximum_output_size(input.len()),
     );
-    let mut enc = FrameEncoder::with_frame_info(frame_info, buffer);
-    enc.write_all(input).unwrap();
-    enc.finish().unwrap()
+    let mut enc = FrameEncoder::with_frame_info(frame_info, buffer)?;
+    enc.write_all(input)?;
+    Ok(enc.finish()?)
 }
 
 /// Decompress all bytes of `input` into a new vec.
 pub fn decompress(input: &[u8]) -> Result<Vec<u8>, Error> {
     let mut de = FrameDecoder::new(input);
-    // Preallocate the Vec with 1.5x the size of input, it may resize but it amortizes enough.
+    // Preallocate the output with 2x the input size, it may resize but it amortizes enough.
     // The upside is that we don't have to worry about DOS attacks, etc..
-    let mut out = Vec::with_capacity(input.len() * 3 / 2);
+    let mut out = Vec::with_capacity(input.len() * 2);
     de.read_to_end(&mut out)?;
     Ok(out)
 }
+
+/// Compress `input` into `output`.
+pub fn compress_into(input: &mut impl Read, output: &mut impl Write) -> Result<(), Error> {
+    compress_into_with(Default::default(), input, output)
+}
+
+/// Compress `input` into `output` with the specified Frame configuration.
+pub fn compress_into_with(
+    frame_info: FrameInfo,
+    input: &mut impl Read,
+    output: &mut impl Write,
```

Comment:
Isn't it common to take `Read` and `Write` by value and the caller needs to make sure that the &mut references implement Write and Read

Reply:
I don't know honestly. I was following `io::copy` signature here, which is semantically equivalent.

```rust
pub fn copy<R: ?Sized, W: ?Sized>(reader: &mut R, writer: &mut W) -> Result<u64> where
    R: Read,
    W: Write,
```

Reply:
Having played with these functions a bit more I think forcing &mut onto the users (like io::copy) is quite unergonomic. Should I switch it like you suggested?

Reply:
Yes, I checked some API's and this seems to be the common case